src/bgw/job.c

```diff
@@ -176,77 +180,152 @@ ts_bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
+	BgwJob *job = NULL;
+	LOCKTAG tmp;
+	LOCKTAG *tag = session_lock_locktag;
+
+	if (tag == NULL)
+		tag = &tmp;
+
+	/* take advisory lock before relation lock */
+	if (!lock_job(bgw_job_id, tuple_lock_mode, session_lock_locktag != NULL, tag, block))
+	{
+		/* return NULL if lock could not be acquired */
+		Assert(!block);
+		return NULL;
+	}
 
-	*job_pp = bgw_job_from_tuple(ti->tuple, sizeof(BgwJob), ti->mctx);
+	init_scan_by_job_id(&iterator, bgw_job_id);
 
-	/*
-	 * Return SCAN_CONTINUE because we check for multiple tuples as an error
-	 * condition.
-	 */
-	return SCAN_CONTINUE;
+	ts_scanner_foreach(&iterator)
+	{
+		HeapTuple heap = ts_scan_iterator_tuple(&iterator);
+		job = bgw_job_from_tuple(heap, sizeof(BgwJob), mctx);
+
+		Assert(num_found == 0);
+		num_found++;
+	}
+
+	return job;
+}
+
+/* Take a share lock on the job for the duration of the txn. This prevents
+ *  the job from being deleted.
+ *
+ *  Returns whether or not the job still exists.
+ */
+bool
+ts_bgw_job_get_share_lock(int32 bgw_job_id, MemoryContext mctx)
+{
+	/* note the mode here is equivalent to FOR SHARE row locks */
+	BgwJob *job = ts_bgw_job_find_with_lock(bgw_job_id, mctx, RowShareLock, NULL, true);
+	if (job != NULL)
+	{
+		pfree(job);
+		return true;
+	}
+	return false;
 }
 
 BgwJob *
 ts_bgw_job_find(int32 bgw_job_id, MemoryContext mctx, bool fail_if_not_found)
 {
-	BgwJob *job_stat = NULL;
+	ScanIterator iterator = ts_scan_iterator_create(BGW_JOB, AccessShareLock, mctx);
+	int num_found = 0;
+	BgwJob *job = NULL;
+
+	init_scan_by_job_id(&iterator, bgw_job_id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		Assert(num_found == 0);
+		job = bgw_job_from_tuple(ts_scan_iterator_tuple(&iterator), sizeof(BgwJob), mctx);
+		num_found++;
+	}
+
+	if (num_found == 0 && fail_if_not_found)
+		elog(ERROR, "job %d not found", bgw_job_id);
+
+	return job;
+	;
+}
+
+static void
+get_job_lock_for_delete(int32 job_id)
+{
+	LOCKTAG tag;
+	bool got_lock;
+
+	/* Try getting an exclusive lock on the tuple in a non-blocking manner. Note this is the
+	 * equivalent of a row-based FOR UPDATE lock */
+	got_lock = lock_job(job_id,
+						AccessExclusiveLock,
+						/* session_lock */ false,
+						&tag,
+						/* block */ false);
+	if (!got_lock)
+	{
+		/* If I couldn't get a lock, try killing the background worker that's running the job.
+		 * This is probably not bulletproof but best-effort is good enough here. */
+		VirtualTransactionId *vxid = GetLockConflicts(&tag, AccessExclusiveLock);
+		PGPROC *proc;
 
-	bgw_job_scan_job_id(bgw_job_id,
-						bgw_job_tuple_found,
-						NULL,
-						&job_stat,
-						mctx,
-						AccessShareLock,
-						fail_if_not_found);
+		if (VirtualTransactionIdIsValid(*vxid))
+		{
+			proc = BackendIdGetProc(vxid->backendId);
+			if (proc != NULL && proc->isBackgroundWorker)
+			{
+				elog(NOTICE,
+					 "cancelling the background worker for job %d (pid %d)",
+					 job_id,
+					 proc->pid);
+				DirectFunctionCall1(pg_cancel_backend, Int32GetDatum(proc->pid));
```

Comment:
Hmmm, on further thought, I'm not sure how comfortable I am with this in the context of multi-transaction BGWs, as it creates more potential for there to be leftover state. _In principle_ we need to be able to handle any BGW just dying, so we should be able to handle it, but since it is such and edge case, it probably will never be tested as well as we like, and if we can avoid entering such states, maybe we should?

Reply:
(The other option would be to have a function to interrupt a BGW [possibly just `kill`], that would also allow the user to detect if the job has quiesced before the job is deleted. A downside would be that there wouldn't be an atomic way to kill-and-delete a job that keeps restarting itself)